File: java/src/main/java/problems1_100/MergeIntervals.java
package problems1_100;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Comparator;
import java.util.List;

enum MergeIntervals {
    DEFAULT {
        @Override
        int[][] merge(int[][] intervals) {
            Arrays.sort(intervals, Comparator.comparingInt(array -> array[0]));
            List<int[]> newIntervals = new ArrayList<>();
            newIntervals.add(intervals[0]);
            for (int i = 1; i < intervals.length; i++) {
                int[] last = newIntervals.get(newIntervals.size() - 1);
                if (last[1] < intervals[i][0]) {
                    newIntervals.add(intervals[i]);
                    continue;
                }
                newIntervals.remove(newIntervals.size() - 1);
                newIntervals.add(
                    new int[]{
                        last[0],
                        Math.max(last[1], intervals[i][1]),
                    }
                );
            }

            int[][] result = new int[newIntervals.size()][2];
            for (int i = 0; i < newIntervals.size(); i++) {
                result[i] = newIntervals.get(i);
            }
            return result;
        }
    };

    abstract int[][] merge(int[][] intervals);
}

File: kotlin/src/main/kotlin/problems1_100/MergeIntervals.kt
package problems1_100

import java.util.Arrays
import kotlin.math.max

enum class MergeIntervals {
    DEFAULT {
        override fun merge(intervals: Array<IntArray>): Array<IntArray> {
            Arrays.sort(intervals, Comparator.comparingInt { array: IntArray -> array[0] })
            val newIntervals = arrayListOf(intervals[0])
            for (i in 1 until intervals.size) {
                val last = newIntervals[newIntervals.size - 1]
                if (last[1] < intervals[i][0]) {
                    newIntervals += intervals[i]
                    continue
                }
                newIntervals.removeAt(newIntervals.size - 1)
                newIntervals +=
                    intArrayOf(
                        last[0],
                        max(last[1].toDouble(), intervals[i][1].toDouble()).toInt(),
                    )
            }

            val result = Array(newIntervals.size) { IntArray(2) }
            for (i in newIntervals.indices) {
                result[i] = newIntervals[i]
            }
            return result
        }
    },
    ;

    abstract fun merge(intervals: Array<IntArray>): Array<IntArray>
}
